fix(codegen/linker): unresolved import 이 bare require("...") 로 emit — module factory 스코프에서 ReferenceError

## 요약

의존성 그래프에서 해석(resolve)되지 못한 import 가 codegen 에서 **bare `require("specifier")`** 호출 표현식으로 그대로 emit 된다.

- **IIFE / ESM / RN factory wrapper** 처럼 `require` 가 스코프에 존재하지 않는 환경에서는 즉시 `ReferenceError: Property 'require' doesn't exist` 로 번들 실행이 실패한다.
- 특히 React Native 타겟에서 `import` 대상이 `.d.ts` 전용 경로이거나 `--external` 지정된 패키지일 때 치명적이다. 모듈 factory 스코프에는 `__r` / `_$$_REQUIRE` 같은 내부 require 만 있을 뿐 bare `require` 는 없기 때문이다.

## 환경

- ZTS: `ba07296` (main)
- Consumer: bungae (React Native 번들러) — ZTS NAPI 경유 번들링
- RN: 0.83.4, react-native-screens 4.23.0, Hermes

## 증상 (bungae 에서 앱 부팅 시)

```
[runtime not ready]: ReferenceError: Property 'require' doesn't exist

/.../react-native-screens/src/components/helpers/prepareHeaderBarButtonItems.ts
  __init
/.../react-native-screens/src/components/ScreenStackHeaderConfig.tsx
  __init
...
```

ZTS 가 뽑은 번들의 해당 모듈 본문 (bungae dev server, line 105518~105522):

```js
var init_react_native_screens_src_components_helpers_prepareHeaderBarButtonItems = __esm({
  ".../prepareHeaderBarButtonItems.ts"() {
    __ns_881_0 = __toESM(require_react_native_index());
    var HeaderBarButtonItem = require("react-native-screens/types").HeaderBarButtonItem;          // ❌
    var HeaderBarButtonItemWithMenu = require("react-native-screens/types").HeaderBarButtonItemWithMenu; // ❌
    ...
  }
});
```

원본 TS:

```ts
// react-native-screens/src/components/helpers/prepareHeaderBarButtonItems.ts
import { Image, processColor } from 'react-native';
import {
  HeaderBarButtonItem,
  HeaderBarButtonItemWithMenu,
} from 'react-native-screens/types';
```

- 첫 번째 `'react-native'` import 는 정상적으로 `require_react_native_index()` 로 rewrite 되었다.
- 두 번째 `'react-native-screens/types'` import 는:
  - 해당 경로는 package.json 에 subpath export 가 없고 실제 파일은 `src/types.tsx` 만 존재 (package.json `types` 는 `lib/typescript/index.d.ts` 전용) → 그래프에서 **resolve 실패**.
  - 심볼 `HeaderBarButtonItem`, `HeaderBarButtonItemWithMenu` 는 원본 TS 에서 **타입 파라미터로만 사용**되며 값으로는 한 번도 참조되지 않는다.
  - 두 조건 모두 충족했는데도 import 가 제거되지 않고, bare `require("react-native-screens/types").<name>` 로 emit 되었다.

## 최소 재현 (ZTS standalone)

```
# /tmp/zts-repro/types.d.ts
export type Foo = { a: string };
export type Bar = { b: number };

# /tmp/zts-repro/entry.ts
import { Foo, Bar } from './types';
export function make(): Foo { return { a: 'x' }; }
export function make2(): Bar { return { b: 1 }; }
```

`--external ./types` 로 resolve 실패 상황을 시뮬레이션:

```
$ zts --bundle /tmp/zts-repro/entry.ts --external ./types --format=esm
// --- entry.ts ---
var Foo = require("./types").Foo;
var Bar = require("./types").Bar;
function make() { return { a: "x" }; }
function make2() { return { b: 1 }; }
export { make, make2 };
```

- `--format=esm` / `--format=iife` 둘 다 동일하게 `require(...)` 로 떨어진다. 브라우저/RN factory wrapper 에서는 ReferenceError.
- `--format=cjs` 는 런타임에 bare `require` 가 존재하므로 에러까지는 안 가지만, 타입 전용 경로면 런타임에 실제 모듈이 없어 throw.
- 참고로 `Foo`, `Bar` 는 순수 TS 타입 전용 심볼인데도 value import 로 살아남아 emit 된다 (아래 2번 참고).

## 원인 분석 (소스 인용)

### 1. `buildRequireRewrites` — unresolved import 을 rewrite 맵에 넣지 않고 skip

`src/bundler/linker/metadata.zig` 584~601:

```zig
pub fn buildRequireRewrites(self: *const Linker, m: *const Module) !std.StringHashMapUnmanaged([]const u8) {
    var require_rewrites: std.StringHashMapUnmanaged([]const u8) = .{};
    const self_idx = m.index.toU32();
    for (m.import_records) |rec| {
        if (rec.resolved.isNone()) {
            // UMD/AMD: external require → factory 매개변수 참조.
            if (rec.is_external and (self.format == .umd or self.format == .amd)) {
                ...
            }
            continue;   // ← ESM / IIFE / CJS / RN factory-wrap 케이스에서 맵에 아무것도 등록 안 됨
        }
        ...
    }
}
```

`resolved.isNone()` 인 경우 UMD/AMD 만 factory 파라미터로 바인딩하고, 그 외 포맷은 그냥 `continue` 한다. 결과적으로 이 specifier 에 대한 rewrite 엔트리가 만들어지지 않는다.

### 2. `emitRequireRewriteOrCall` — rewrite 엔트리가 없으면 `require(...)` 를 통째로 찍음

`src/codegen/codegen.zig` 2114~2124:

```zig
fn emitRequireRewriteOrCall(self: *Codegen, source: ast_mod.NodeIndex) !bool {
    if (self.resolveRequireRewrite(source)) |req_var| {
        try self.emitRewriteValue(req_var);
        return true;
    }
    try self.write("require(");
    try self.emitNode(source);
    try self.writeByte(')');
    return false;
}
```

맵에 엔트리가 없으니 `resolveRequireRewrite` 는 `null` → 바로 bare `require("specifier")` 를 써버린다. 이 함수가 호출되는 상위 경로(`emitImportCJS` 라인 2910~2917)도 specifier 를 쭉 돌며 그대로 붙인다.

### 3. (부가) Type-only 스펙 elision 부재

`src/parser/module.zig` 의 `parseImportSpecifier` 는 inline `type` modifier 를 감지해 `import_specifier` 노드의 `flags = is_type_only (0/1)` 에 저장한다 (라인 404, 474). 하지만 codegen / transformer / linker 어디서도 이 플래그를 참조하지 않는다:

```
$ grep -rn 'is_type_only\|type_only' src/codegen/ src/transformer/ src/bundler/
src/transformer/transformer/class_decorator.zig:609:    const type_only_mask = ast_mod.PropertyFlags.is_abstract | ...
```

→ class 필드 쪽만 쓰이고 `import_specifier` flags 는 dead. 또한 TS 관점의 "value import 인데 값으로 단 한 번도 참조되지 않은" 심볼 elision (TypeScript `isolatedModules`/`verbatimModuleSyntax=false` 동작) 도 구현되어 있지 않다. 본 이슈 케이스는 이 두 번째(사용처 없음) 케이스에 해당한다.

## 왜 문제인가

- **IIFE / ESM / 브라우저 / React Native factory wrapper** 스코프에 `require` 식별자는 존재하지 않는다. bare `require(...)` 를 남기는 순간 invalid JS → ReferenceError.
- 일반적으로 번들러가 unresolved / external import 를 만났을 때 취할 수 있는 선택지:
  1. 빌드 에러 (default)
  2. ESM 그대로 유지 (`import { X } from 'specifier'` — esbuild `--external` 의 ESM 동작)
  3. format 에 맞는 runtime-safe 스텁 (`throw new Error("Cannot find module '…'")`)
  4. 해당 specifier 를 import 한 심볼이 **값으로 한 번도 참조되지 않으면** import 문 자체를 elide
- 현재 ZTS 는 이 네 가지 중 어느 것도 아닌 "bare `require(...)` 를 일단 적고 본다" 전략인데, 이는 가장 피해야 할 선택이다.

## 제안하는 수정 방향

### 핵심 수정 — bare require emit 제거

`linker/metadata.zig:buildRequireRewrites` 또는 `codegen.zig:emitRequireRewriteOrCall` 에서 unresolved import 에 대한 format 별 처리를 분기:

| format            | 처리                                                                             |
|-------------------|----------------------------------------------------------------------------------|
| `cjs`             | bare `require("spec")` 유지 가능 (Node CJS 스코프엔 require 존재)                |
| `esm`             | `import { X, Y } from "spec"` 원형 유지 (external 의미)                           |
| `iife`            | 빌드 에러 (`require` 주입 불가)                                                   |
| `umd` / `amd`     | 이미 구현됨 (factory 파라미터 바인딩)                                            |
| RN consumer (wrap) | factory 내부에서 호출 가능한 내부 심볼(`__r(dep_idx)` 등) 로 rewrite 필요. NAPI 옵션으로 consumer 가 rewrite 전략 주입할 수 있어야 함 |

추가 제안:
- `strict_unresolved: bool` 같은 옵션으로 default 를 "unresolved → error" 로 두고, consumer(bungae) 가 명시적으로 "warn + keep original import" 를 선택하게.

### 부가 수정 — type-only specifier elision

1. `import_specifier.flags & 1 == 1` 인 specifier 는 emit 단계에서 skip.  
   위치: `src/codegen/codegen.zig:2910-2917` (named specifier 루프) 및 `emitImportCJS`.
2. `import { X, Y } from 'mod'` 전체가 type-only modifier (`import type { ... }`) 인 구문의 elision 여부도 함께 점검.
3. 더 나아가 semantic analyzer 에서 `reference_count == 0` 이고 value 로 참조되지 않은 import 심볼은 transformer 단계에서 drop (TS 기본 동작 — `verbatimModuleSyntax=false`).

두 수정이 합쳐지면 본 이슈 케이스는:
- (부가) type-only elision 으로 `import { Foo, Bar } from './types'` 가 AST 에서 제거
- 혹은 (핵심) 제거되지 않더라도 bare `require` 대신 format 에 맞는 안전한 emit

어느 쪽이든 runtime crash 는 발생하지 않게 된다.

## 참고

- bungae 쪽 영향: 이 버그 때문에 `react-native-screens` 를 사용하는 RN 앱은 ZTS 번들로 부팅 자체가 불가능 (ExpoApp + expo-router 플로우에서 재현 확인).
- 최근 `require.context` 관련 Phase 3/4 변경(커밋 5a1f35d, c631da2, 40e1f2e 등) 에서 emit 파이프라인에 손이 많이 갔는데, 거기서 도입됐을 가능성도 있어 git blame 으로 추적해볼 가치가 있다.

--- a/src/bundler/bundler.zig
+++ b/src/bundler/bundler.zig
@@ -361,6 +361,28 @@ pub const BundleResult = struct {
     }
 };
 
+/// BundlerDiagnostic 을 allocator-owned OwnedDiagnostic 으로 deep copy 해 `dest` 에 채운다.
+/// `filled` 는 이미 채워진 item 수 — 루프 내부에서 매 entry 성공 후 증가한다 (호출자의
+/// errdefer 가 부분 할당분을 정확히 해제할 수 있도록).
+fn copyDiagnostics(
+    allocator: std.mem.Allocator,
+    dest: []BundleResult.OwnedDiagnostic,
+    src: []const types.BundlerDiagnostic,
+    filled: *usize,
+) !void {
+    for (src) |d| {
+        dest[filled.*] = .{
+            .code = d.code,
+            .severity = d.severity,
+            .message = try allocator.dupe(u8, d.message),
+            .file_path = try allocator.dupe(u8, d.file_path),
+            .step = d.step,
+            .suggestion = if (d.suggestion) |s| try allocator.dupe(u8, s) else null,
+        };
+        filled.* += 1;
+    }
+}
+
 pub const Bundler = struct {
     allocator: std.mem.Allocator,
     options: BundleOptions,
@@ -1003,9 +1025,12 @@ pub const Bundler = struct {
         // 이 `t_graph/t_link/t_shake/t_emit` 은 `BundleResult.timings` 를 채워
         // NAPI `WatchRebuildEvent.phaseDurations` 로 노출 (HMR 관측성).
 
-        // 4. 진단 메시지 deep copy (graph.deinit 후에도 문자열 유효하도록)
-        const diagnostics: ?[]BundleResult.OwnedDiagnostic = if (graph.diagnostics.items.len > 0) blk: {
-            const diags = try self.allocator.alloc(BundleResult.OwnedDiagnostic, graph.diagnostics.items.len);
+        // 4. 진단 메시지 deep copy (graph.deinit 후에도 문자열 유효하도록).
+        // graph.diagnostics + linker.fatal_diagnostics (IIFE unresolved 등, #1791) 병합.
+        const link_diag_len = if (linker) |*l| l.fatal_diagnostics.items.len else 0;
+        const diagnostics: ?[]BundleResult.OwnedDiagnostic = if (graph.diagnostics.items.len > 0 or link_diag_len > 0) blk: {
+            const total = graph.diagnostics.items.len + link_diag_len;
+            const diags = try self.allocator.alloc(BundleResult.OwnedDiagnostic, total);
             errdefer self.allocator.free(diags);
             // M1 수정: 부분 할당 후 OOM 시 이미 복사한 문자열 해제
             var filled: usize = 0;
@@ -1014,16 +1039,9 @@ pub const Bundler = struct {
                 self.allocator.free(d.file_path);
                 if (d.suggestion) |s| self.allocator.free(s);
             };
-            for (graph.diagnostics.items, 0..) |d, i| {
-                diags[i] = .{
-                    .code = d.code,
-                    .severity = d.severity,
-                    .message = try self.allocator.dupe(u8, d.message),
-                    .file_path = try self.allocator.dupe(u8, d.file_path),
-                    .step = d.step,
-                    .suggestion = if (d.suggestion) |s| try self.allocator.dupe(u8, s) else null,
-                };
-                filled = i + 1;
+            try copyDiagnostics(self.allocator, diags, graph.diagnostics.items, &filled);
+            if (linker) |*l| {
+                try copyDiagnostics(self.allocator, diags, l.fatal_diagnostics.items, &filled);
             }
             break :blk diags;
         } else null;

--- a/src/bundler/emitter.zig
+++ b/src/bundler/emitter.zig
@@ -1047,6 +1047,20 @@ pub fn emitModule(
         if (override_syms) |syms| {
             md.symbol_ids = syms;
         }
+        // #1791 IIFE unresolved build-time diag 를 linker 의 공용 `fatal_diagnostics`
+        // 로 소유권 이전. mutex 는 공용 리스트의 append 만 보호 — `md.pending_diagnostics`
+        // 자체는 per-module 이라 경쟁 없음. `@constCast` 는 `ns_cache_mutex` (linker.zig:1877)
+        // 와 동일 관행. free 후 필드를 비워 md.deinit 의 double-free 방지.
+        if (md.pending_diagnostics.len > 0) {
+            const linker_mut = @constCast(l);
+            linker_mut.diagnostics_mutex.lock();
+            defer linker_mut.diagnostics_mutex.unlock();
+            for (md.pending_diagnostics) |d| {
+                try linker_mut.fatal_diagnostics.append(l.allocator, d);
+            }
+            l.allocator.free(md.pending_diagnostics);
+            md.pending_diagnostics = &.{};
+        }
         // statement-level tree-shaking: StmtInfo 기반 도달성 분석으로 미사용 statement 제거.
         // rolldown 방식: 심볼 인덱스로 추적하여 linker rename 후에도 정확한 판정.
         //

--- a/src/bundler/linker.zig
+++ b/src/bundler/linker.zig
@@ -203,6 +203,11 @@ pub const LinkingMetadata = struct {
     /// 기본값 true 는 현재 모든 생성 지점이 번들러(`buildMetadataForAst` 계열)라는 사실에
     /// 의존한다. 비-번들러 생성 지점을 추가할 때는 반드시 false 를 명시할 것.
     is_bundle_context: bool = true,
+    /// #1791: build-time 에러 (e.g. IIFE 포맷 + unresolved import). `buildMetadataForAst`
+    /// 가 병렬 호출되므로 직접 `linker.diagnostics` 에 append 하면 race. 대신 이 리스트에
+    /// 쌓고 emitter 가 serial 경로에서 `linker.fatal_diagnostics` 로 flush.
+    /// item.message 는 allocator 소유 — flush 시 소유권 이전, deinit 에서 free.
+    pending_diagnostics: []const BundlerDiagnostic = &.{},
     allocator: std.mem.Allocator,
 
     pub const NsMemberRewrites = struct {
@@ -283,6 +288,12 @@ pub const LinkingMetadata = struct {
             for (vars) |v| self.allocator.free(v);
             self.allocator.free(vars);
         }
+        // Ownership: 정상 경로는 emitter 가 message 소유권을 `linker.fatal_diagnostics`
+        // 로 이전한 후 slice 를 비우므로 여기서 no-op. flush 전 에러 경로에서만 해제.
+        if (self.pending_diagnostics.len > 0) {
+            for (self.pending_diagnostics) |d| self.allocator.free(d.message);
+            self.allocator.free(self.pending_diagnostics);
+        }
     }
 };
 
@@ -318,6 +329,14 @@ pub const Linker = struct {
     resolved_bindings: std.AutoHashMap(BindingKey, ResolvedBinding),
 
     diagnostics: std.ArrayList(BundlerDiagnostic),
+    /// #1791 사용자 노출용 치명 진단 (예: IIFE 포맷에서 unresolved import).
+    /// 기존 `diagnostics` 는 내부/테스트 전용 — bundler 가 BundleResult 로 wire 하지
+    /// 않는다. 이 필드로 들어온 항목만 사용자에게 `build error` 로 노출된다.
+    /// message 는 allocator 소유 (allocPrint) — linker.deinit 에서 일괄 해제.
+    fatal_diagnostics: std.ArrayList(BundlerDiagnostic) = .empty,
+    /// #1791 emitter 가 `emitModuleThread` 로 병렬 emit 중 `LinkingMetadata.pending_diagnostics`
+    /// 를 linker 의 `fatal_diagnostics` 버퍼로 flush. 병렬 append 보호.
+    diagnostics_mutex: std.Thread.Mutex = .{},
 
     /// semantic.Symbol.canonical_name 슬라이스의 backing 저장소. linker가 소유 —
     /// deinit에서 일괄 해제. AliasTable.canonical_name은 caller-owned (별도 모델).
@@ -470,6 +489,9 @@ pub const Linker = struct {
         var nic_it = self.ns_inline_cache.valueIterator();
         while (nic_it.next()) |v| self.allocator.free(v.*);
         self.ns_inline_cache.deinit(self.allocator);
+        // #1791 fatal diag message 해제 (allocPrint owned)
+        for (self.fatal_diagnostics.items) |d| self.allocator.free(d.message);
+        self.fatal_diagnostics.deinit(self.allocator);
         self.diagnostics.deinit(self.allocator);
     }
 

--- a/src/bundler/linker/metadata.zig
+++ b/src/bundler/linker/metadata.zig
@@ -192,6 +192,18 @@ pub fn buildMetadataForAst(
         ns_var_list.deinit(self.allocator);
     }
 
+    // #1791 IIFE unresolved build-time diag. 정상 경로에서는 emitter 가 serial 로
+    // `linker.fatal_diagnostics` 에 flush. item.message 는 allocator 소유.
+    var pending_diags: std.ArrayListUnmanaged(@import("../types.zig").BundlerDiagnostic) = .empty;
+    errdefer {
+        for (pending_diags.items) |d| self.allocator.free(d.message);
+        pending_diags.deinit(self.allocator);
+    }
+    // IIFE dedupe 맵은 실제 IIFE 포맷 + unresolved 발생 시에만 lazy init — ESM/CJS
+    // 빌드 (실측 99%+ 모듈) 에서 해시맵 backing alloc 비용을 피한다.
+    var iife_diag_seen: ?std.StringHashMap(void) = null;
+    defer if (iife_diag_seen) |*seen_map| seen_map.deinit();
+
     // namespace import / inline object 캐시는 linker 전역 필드(`self.ns_export_cache`,
     // `self.ns_inline_cache`) 로 이동 — 같은 target 을 여러 모듈이 namespace import 해도
     // collectExportsRecursive DFS 를 한 번만 수행하도록 공유 (#1734).
@@ -242,8 +254,34 @@ pub fn buildMetadataForAst(
                             try preamble.write(ib.imported_name);
                             try preamble.write(";\n");
                         }
+                    } else if (self.format == .iife and !ib.isSynthetic()) {
+                        // #1791 IIFE: unresolved external 은 의미 자체가 성립 안 함
+                        // (factory 스코프에 require 없음, top-level import 도 불가).
+                        // build-time 에러로 pending_diagnostics 에 쌓고 emitter 가 flush.
+                        // specifier 단위로 dedupe — 같은 spec 의 binding 여러 개에 대해 한 번만.
+                        if (iife_diag_seen == null) iife_diag_seen = std.StringHashMap(void).init(self.allocator);
+                        const seen_gop = try iife_diag_seen.?.getOrPut(rec.specifier);
+                        if (!seen_gop.found_existing) {
+                            const msg = try std.fmt.allocPrint(
+                                self.allocator,
+                                "unresolved import \"{s}\" cannot be emitted in IIFE format (no require/import available in factory scope)",
+                                .{rec.specifier},
+                            );
+                            errdefer self.allocator.free(msg);
+                            try pending_diags.append(self.allocator, .{
+                                .code = .unresolved_import,
+                                .severity = .@"error",
+                                .message = msg,
+                                .file_path = m.path,
+                                .span = rec.span,
+                                .step = .resolve,
+                                .suggestion = null,
+                            });
+                        }
                     } else {
-                        // ESM/CJS/IIFE: require() preamble 생성
+                        // ESM/CJS: require() preamble 생성. ESM 출력에서도 esbuild 호환으로
+                        // require() 를 유지 — Node.js 가 `import` 없는 출력을 CJS 로 파싱하여
+                        // `var X; var X;` 재선언을 허용하게 한다 (`emitter.zig` 상단 주석 참조).
                         if (is_synthetic_esm) {
                             try preamble.writeUnresolvedRequireAssignOnly(preamble_name, rec.specifier, ib.imported_name, ib.kind == .namespace);
                         } else {
@@ -561,6 +599,12 @@ pub fn buildMetadataForAst(
     else
         null;
 
+    // #1791 IIFE unresolved 진단 소유권 이전
+    const pending_diags_slice: []const @import("../types.zig").BundlerDiagnostic = if (pending_diags.items.len > 0)
+        try pending_diags.toOwnedSlice(self.allocator)
+    else
+        &.{};
+
     return .{
         .skip_nodes = skip_nodes,
         .renames = renames,
@@ -575,6 +619,7 @@ pub fn buildMetadataForAst(
         .export_getter_overrides = export_getter_overrides,
         .owned_rename_values = owned_nested_renames,
         .dev_ns_vars = dev_ns_vars,
+        .pending_diagnostics = pending_diags_slice,
         .allocator = self.allocator,
     };
 }

